Add support for updating eMMC boot partitions

BastianStender opened the pull request:
This enables atomic updates of eMMC boot partitions for bootloaders:

- read active boot partition
- empty the inactive boot partition
- set inactive boot partition rw
- write bootloader to inactive partition
- set inactive boot partition ro
- set inactive boot partition active
- check that the previous inactive boot partition is active now

Also the docs were restructured and now reflect the first bootloader update possibility via RAUC.

jluebbe commented on include/emmc.h:
```diff
@@ -0,0 +1,111 @@
+#define EXT_CSD_BOOT_CFG_ACC			(0x07)
+
+/* From kernel linux/major.h */
+#define MMC_BLOCK_MAJOR				179
```
This is unused.

jluebbe commented on src/emmc.c:
```diff
@@ -0,0 +1,189 @@
+#include "update_handler.h"
+
+
+int r_emmc_read_extcsd(int fd, __u8 *extcsd) {
```
This can be static. Remove it from the header es well.

jluebbe commented on src/emmc.c:
```diff
@@ -0,0 +1,189 @@
+	return res;
+}
+
+gboolean r_emmc_force_part_write(const gchar *device, gchar value, GError **error) {
```
This seems to be internal as well, so make it static.

jluebbe commented on src/update_handler.c:
```diff
@@ -1034,6 +1083,160 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+	GError *ierror = NULL;
+	RaucSlot *part_slot = NULL;
+
+	g_debug("img_to_boot_emmc_handler");
```
Remove this.

jluebbe commented on src/update_handler.c:
```diff
@@ -56,6 +57,54 @@ static GOutputStream* open_slot_device(RaucSlot *slot, int *fd, GError **error)
 	return outstream;
 }
 
+static gboolean empty_slot(RaucSlot *slot, GError **error)
```
Maybe rename this to "clear_slot"?

jluebbe commented on src/update_handler.c:
```diff
@@ -56,6 +57,54 @@ static GOutputStream* open_slot_device(RaucSlot *slot, int *fd, GError **error)
+	GError *ierror = NULL;
+	GInputStream *instream;
+	GOutputStream *outstream = NULL;
+	GFile *zerosrc = g_file_new_for_path("/dev/zero");
```
Using /dev/zero from C seems strange. We could just write a zeroed page from memory until the device is full.

BastianStender commented:
Pushed:
- squashed @jluebbe's doc fixup
- removed `MMC_BLOCK_MAJOR`, included `linux/major.h` instead (needed to make `MMC_IOC_CMD` macro work)
- made `r_emmc_read_extcsd()`, `r_emmc_write_extcsd()`, `r_emmc_force_part_write()` static, removed from header
- renamed `empty_slot()` to `clear_slot()`
- `clear_slot()`: write zero buffer instead of GInputStream from `/dev/zero`
- unnecessary debug output removed

BastianStender commented:
Pushed:
- `clear_slot()`: fix indent (uncrustified)

ejoerns commented on docs/advanced.rst:
```diff
@@ -517,39 +519,66 @@ non-bootable system caused by an error or power-loss during the update.
   version shipped in the (new) rootfs. In that case:
 
-  * Disable the old rootfs slot and update the bootloader.
+  * Disable the old rootfs slot and update the bootloader via
```
Mh... not sure if this isn't a bit broken. Doesn't write-slot require you to define the slot in the system.conf despite we intentionally do not want to handle this slot with RAUC?..

BastianStender replied:
True. Any suggestions? Either define the slot in the system.conf without providing updates in bundles, or write the bootloader without `rauc slot-write`.

jluebbe commented on src/emmc.c:
```diff
@@ -0,0 +1,190 @@
+static int r_emmc_write_extcsd(int fd, __u8 index, __u8 value) {
+	struct mmc_ioc_cmd cmd;
+
+	memset(&cmd, 0, sizeof(cmd));
```
You can use struct mmc_ioc_cmd cmd = {}; instead of the memset.

jluebbe commented on src/emmc.c:
```diff
@@ -0,0 +1,190 @@
+	g_assert(value == '0' || value == '1');
+
+	/* extract mmcblkXbootY from /dev/mmcblkXbootY */
+	device_split = g_strsplit(device, "/", 3);
```
You could use g_path_get_basename(), which never fails.

jluebbe commented on src/emmc.c:
```diff
@@ -0,0 +1,190 @@
+
+	if (!g_file_test(sysfs_path, G_FILE_TEST_EXISTS)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not generate sysfs path from device: %s does not exist",
```
The problem is not generating the path here, but the configured block devices doesn't have the force_ro attribute.

ejoerns commented on src/emmc.c:
```diff
@@ -0,0 +1,190 @@
+gboolean r_emmc_read_bootpart(const gchar *device, gint *bootpart_active, GError **error) {
+	int ret;
+	gboolean res = FALSE;
+	__u8 extcsd[512];
```
maybe prefer using guint8 here as we normally use glib types.

jluebbe commented on src/update_handler.c:
```diff
@@ -56,6 +59,51 @@ static GOutputStream* open_slot_device(RaucSlot *slot, int *fd, GError **error)
+		 */
+		if (write_count == -1 &&
+		    !g_error_matches(ierror, G_IO_ERROR, G_IO_ERROR_NO_SPACE)) {
+
```
Remove empty line.

ejoerns commented on src/emmc.c:
```diff
@@ -0,0 +1,190 @@
+	int active_partition = -1;
+
+	fd = open(device, O_RDONLY);
+
```
please remove newline here and in similar cases (in some you already did that).

jluebbe commented on src/update_handler.c:
```diff
@@ -56,6 +59,51 @@ static GOutputStream* open_slot_device(RaucSlot *slot, int *fd, GError **error)
+	return res;
+}
+
+
```
Remove empty line.

jluebbe commented on src/update_handler.c:
```diff
@@ -1034,6 +1082,158 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+		}
+	}
+
+	/* read active boot partition vom ext_csd */
```
"from" instead of "vom"

jluebbe commented on src/update_handler.c:
```diff
@@ -1034,6 +1082,158 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+			"%sboot%d",
+			dest_slot->device,
+			INACTIVE_BOOT_PARTITION(part_active));
+
```
Remove empty line.

ejoerns commented on src/emmc.c:
```diff
@@ -0,0 +1,190 @@
+
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"opening output device failed: %s", strerror(errno));
```
use `g_strerror()` here

BastianStender replied:
Mh, okay. `strerror()` is used in other places too, e.g. [here](https://github.com/rauc/rauc/blob/master/src/update_handler.c#L41) and [here](https://github.com/rauc/rauc/blob/master/src/update_handler.c#L68). More patch opportunities :)

BastianStender commented:
Pushed changes:
- use glib types where appropriate
- remove empty lines where appropriate
- `r_emmc_read_extcsd()`: expect `guint8 extcsd[512]` as argument to allow size checking by compiler
- `r_emmc_read_bootpart()`/`r_emmc_read_bootpart()`: mention eMMC device in error messages
- `r_emmc_read_bootpart()`/`r_emmc_read_bootpart()`: close file only if `fd > 0`
- docs/advanced: do not recommend `rauc write-slot` if RAUC should not handle a slot in "old system with a new bootloader" scenario

BastianStender commented:
Pushed changes:
- `r_emmc_read_extcsd`: simplify `mmc_ioc_cmd` struct init with `{}`
- `clear_slot()`: use shorter struct init `{}` instead of `{ 0 }`
- `r_emmc_read_bootpart()`/`r_emmc_read_bootpart()`: close file only if `fd >= 0`
- src/emmc.c: include `config.h`

BastianStender commented:
Pushed changes:
- `r_emmc_force_part_write()`: simplify basename extraction via `g_path_get_basename`

BastianStender commented:
Pushed changes:
- src/emmc.c: drop const keyword as it is not really needed here and complicates freeing; add missing `g_free()`

BastianStender commented:
Ah, thanks for that :)

BastianStender commented:
Pushed changes:
- squashed @jluebbe's const fixup

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+	}
+
+	/* read active boot partition from ext_csd */
+	ret = r_emmc_read_bootpart(dest_slot->device, &part_active, &ierror);
```
return type of `r_emmc_read_bootpart()` is gboolean, thus do not assign it to a `gint` variable. Use `res` here instead, too, which would also save a line of code (`res = FALSE`).

BastianStender replied:
Yes, that's refactor fallout, will fix.

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+	/* create a temporary RaucSlot with the actual partition as device */
+	part_slot = g_new0(RaucSlot, 1);
+
+	/* determine active boot partition and set the other to be written */
```
misleading as we determined the active one already. Maybe the prior comment already describes both and you could just remove the empty line between them.

BastianStender replied:
Ok.

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+	g_message("Found active eMMC boot partition %sboot%d",
+			dest_slot->device, part_active);
+
+	/* create a temporary RaucSlot with the actual partition as device */
```
maybe s/actual partition/actual (currently inactive) boot partition/ ?

BastianStender replied:
Ok.

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+	}
+
+	/* clear block device */
+	g_message("Clearing slot device %s", part_slot->device);
```
The messages are a bit inconsistent when speaking about `part_slot->device` as either *slot device* or *slot device partition*.

BastianStender replied:
Ok, will clarify.

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+	 * Read explicitly from root device (this forces another kernel
+	 * partition switch and should trigger the ext_csd bug more reliably.
+	 */
+	ret = r_emmc_read_bootpart(dest_slot->device, &part_active_after, &ierror);
```
See note about `gboolean` above.

BastianStender replied:
True.

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+		goto out;
+	}
+
+	/* sanity check: read active boot partition from ext_csd
```
s/read/readd/

BastianStender replied:
What does this mean?

ejoerns replied:
Skip that :D seems I confused while scrolling through the code and assumed this to be a re-add from 'adding again' but its indeed a simple 'read' ;)

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+	/* sanity check: read active boot partition from ext_csd
+	 *
+	 * Read explicitly from root device (this forces another kernel
+	 * partition switch and should trigger the ext_csd bug more reliably.
```
close bracket here

BastianStender replied:
Ok.

ejoerns commented on src/emmc.c:
```diff
@@ -0,0 +1,178 @@
+	gchar *sysfs_path = NULL;
+	FILE *f = NULL;
+
+	g_assert(value == '0' || value == '1');
```
Please use `g_return_{val_,}if_fail()` for checking input args of functions. Also consider checking other inputs, such as GError, too.

BastianStender replied:
I see the point. Is this something you are trying to enforce everywhere now? There are different variants of g_assert used for argument checking in the very same file, e.g. [here](https://github.com/rauc/rauc/blob/master/src/update_handler.c#L557).

ejoerns replied:
Yes, we started with `g_assert` in some points but we would like to stick to glib convention and use g_return.. functions here. We tried to enforce this in the last patches added (while not being sure if we succeeded). The intention is to fix this with all newly touched code and later on also in existing code. Anyway, there a still a bunch of functions where we don't do any checks at all. Good patch opportunities.. ;)

ejoerns commented on include/emmc.h:
```diff
@@ -0,0 +1,76 @@
+ * given variable.
+ *
+ * @param device eMMC /dev path (/dev/mmcblkX)
+ * @param bootpart_active will contain the active boot partition (0 or 1)
```
If I get it right, bootpart_active may also contain value `6` when the user area was enabled for boot

BastianStender replied:
Yes, the docstring is wrong.

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+		res = FALSE;
+		goto out;
+	}
+	g_message("Found active eMMC boot partition %sboot%d",
```
This printout will be wrong if `part_active` is 6.

BastianStender replied:
Yes, that's misleading. Will fix it.

ejoerns commented on src/update_handler.c:
```diff
@@ -1034,6 +1080,157 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
+	part_slot->device = g_strdup_printf(
+			"%sboot%d",
+			dest_slot->device,
+			INACTIVE_BOOT_PARTITION(part_active));
```
This will silently switch booting from user partittion to boot1. Might be intended but should at least get an info message, what do you think?

BastianStender replied:
Yes, it's intended, will add comment + appropriate message

BastianStender commented:
> One genereal concern is that the img_to_boot_emmc_handler looks a bit too verbose compared to the others. Maybe we should consider moving some of the messages to debug level?

Yep, will do.

> One last thing to mention in general is that the usage of error class and number is a bit too untargeted imho. You use the tuple R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED for every single error which basically is not what the error classes and numbers are designed for. Might it be worth adding an eMMC error class here? For defining further R_UPDATE errors, the errors might be a bit too specific
> 
> I could image something like
> R_EMMC_EXTCSD_{READ,WRITE}_FAILED, R_EMMC_ERROR, ...
> 
> @jluebbe what do you think about the error case?

We discussed this internally. If error handling is really used anywhere various functions need to be adapter. So we will leave it like this.

BastianStender commented:
Pushed changes as fixup:
- `r_emmc_read_bootpart()`: document values for "user partition" and "disabled"
- `r_emmc_read_bootpart()`: replace `g_assert()` with `g_return_val_if_fail()`
- `img_to_boot_emmc_handler()`: use correct return types and simplify
- `img_to_boot_emmc_handler()`: fix messages about user/disabled boot partition
- `img_to_boot_emmc_handler()`: make comments about device/partition more comprehensive
- `img_to_boot_emmc_handler()`: make it less verbose (rw/ro mode, toggle message are now debug)

Also: rebased on current master

BastianStender commented:
Pushed changes:
- `img_to_boot_emmc_handler()`: fix comment copy/paste error
- squashed
